### PR TITLE
fix for boolean type conversion

### PR DESCRIPTION
**Pull request by akurdyukov**
Fix for #77

### CODE REVIEW

**Review by bryzgaloff — CHANGES_REQUESTED** [review]
Hi @akurdyukov and thank you for your contribution! Please excuse me for a late reply.

I have added several comments. However, I believe the solution to this should be more fundamental: we should hard-code `clickhouse_driver.Client`'s signature into the plugin's code, so that the code fully controls all of the arguments (in contrast to a pretty faulty `**connection_kwargs` approach) and converts them to proper types. This will make the behaviour more predictable.

In other words, I believe there should be a preliminary PR which replaces `Client(**…)` with `Client(host=…, …)` following its signature from `clickhouse-driver` here:
https://github.com/bryzgaloff/airflow-clickhouse-plugin/blob/c7426a1195e1c66da7e1dc2aed2d04eca6db0d61/src/airflow_clickhouse_plugin/hooks/clickhouse.py#L54

And then the PR which you currently work on will easily implement handling this specific str-to-bool case.

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py [hunk 1]
```diff
@@ -90,6 +90,21 @@ def execute(
...
+    """Convert a string representation of truth to true (1) or false (0).
+    True values are 'y', 'yes', 't', 'true', 'on', and '1'; false values
+    are 'n', 'no', 'f', 'false', 'off', and '0'.  Raises ValueError if
+    'val' is anything else.
+    """
+    val = val.lower()
+    if val in ('y', 'yes', 't', 'true', 'on', '1'):
+        return 1
+    elif val in ('n', 'no', 'f', 'false', 'off', '0'):
+        return 0
```
[the remark this answers is not in the document]
Comment (akurdyukov) [reply src/airflow_clickhouse_plugin/hooks/clickhouse.py:103]: Just a generic list of possible bool values people use

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py [hunk 2]
```diff
@@ -105,6 +120,10 @@ def conn_to_kwargs(conn: Connection, database: t.Optional[str]) -> t.Dict[str, t
         connection_kwargs.update(database=database)
     elif conn.schema:
         connection_kwargs.update(database=conn.schema)
+    # converting types for flags
+    for key in ('secure', 'verify'):
+        if key in connection_kwargs:
+            connection_kwargs[key] = strtobool(connection_kwargs[key])
```
[the remark this answers is not in the document]
Comment (akurdyukov) [reply src/airflow_clickhouse_plugin/hooks/clickhouse.py:126]: Yes, you're right, this will fail if Airflow UI is used with correct JSON. Need to check if flags are bool already

**Comment by akurdyukov** [comment]
Thanks for the review! I fixed most of review comments.

Regarding the first one about the method of passing arguments to `clickhouse_driver.Client` - currently there's 23 arguments, most of them are optionals. So, minimal boilerplate version should use something like Pydantic. And it looks like a little overkill to me. What do you think?

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py [hunk 3]
```diff
@@ -90,19 +90,14 @@ def execute(
...
-        return 0
+def strtobool(str_value: str) -> bool:
+    str_value = str_value.lower()
+    if str_value in ('y', 'yes', 't', 'true', 'on', '1'):
+        return True
+    elif str_value in ('n', 'no', 'f', 'false', 'off', '0'):
+        return False
     else:
-        raise ValueError("invalid truth value %r" % (val,))
+        raise ValueError(f'invalid truth value {str_value!r}')
```
Comment (bryzgaloff) [inline src/airflow_clickhouse_plugin/hooks/clickhouse.py:100]: `else` is redundant here

File: tests/unit/hooks/test_strtobool.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        self.assertTrue(strtobool('true'))
+
+    def test_correct_one(self):
+        self.assertTrue(strtobool('1'))
+
+    def test_correct_false(self):
+        self.assertFalse(strtobool('false'))
+
+    def test_correct_zero(self):
+        self.assertFalse(strtobool('0'))
```
Comment (bryzgaloff) [inline tests/unit/hooks/test_strtobool.py:17]: A quick best practice comment: the tests must cover all the supported input values, not only a few. A full coverage is required. However, this is a boilerplate code: you may use `self.subTest` functionality to check all truthy and all falsy values in a loop instead of creating a test per value.

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py (hunk 3 above)
Comment (bryzgaloff) [inline src/airflow_clickhouse_plugin/hooks/clickhouse.py:100]: Why "truth" value btw? What if it was intended to be a falsy one? :)

```suggestion
        raise ValueError(f'unsupported value: {str_value!r}')
```

**Review by bryzgaloff — CHANGES_REQUESTED** [review]

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bryzgaloff) [reply src/airflow_clickhouse_plugin/hooks/clickhouse.py]: Quite biased. Instead let's support only `true`/`false` options. It is impossible to guess what a random user may consider to be a true/false str representation.

**Comment by bryzgaloff** [comment]
> Regarding the first one about the method of passing arguments to `clickhouse_driver.Client` - currently there's 23 arguments, most of them are optionals. So, minimal boilerplate version should use something like Pydantic. And it looks like a little overkill to me. What do you think?

Hi @akurdyukov, yes, definitely using an external library is an overkill here. I suggest to hardcode all the arguments: `clickhouse-driver` is not expected to change them often, so we may accept new PRs once the arguments change in the underlying library.

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py (hunk 3 above)
Comment (akurdyukov) [reply src/airflow_clickhouse_plugin/hooks/clickhouse.py]: fixed

File: tests/unit/hooks/test_strtobool.py (hunk 4 above)
Comment (akurdyukov) [reply tests/unit/hooks/test_strtobool.py]: ok, fixed

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py (hunk 3 above)
Comment (akurdyukov) [reply src/airflow_clickhouse_plugin/hooks/clickhouse.py]: Sure, fixed

File: src/airflow_clickhouse_plugin/hooks/clickhouse.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akurdyukov) [reply src/airflow_clickhouse_plugin/hooks/clickhouse.py]: Alright, only 'true', 'True', 'fase', 'False' supported
